File: src/cavebot/vision/capture.py
"""OBS Virtual Camera frame reader using OpenCV."""

from __future__ import annotations

import cv2
import numpy as np


class FrameCapture:
    def __init__(self, camera_index: int = 0):
        self._camera_index = camera_index
        self._cap: cv2.VideoCapture | None = None

    def open(self) -> bool:
        self._cap = cv2.VideoCapture(self._camera_index)
        return self._cap.isOpened()

    def read(self) -> np.ndarray | None:
        if self._cap is None or not self._cap.isOpened():
            return None
        ret, frame = self._cap.read()
        if not ret:
            return None
        return frame

    def crop_region(self, frame: np.ndarray, region: list[int]) -> np.ndarray:
        x, y, w, h = region
        return frame[y : y + h, x : x + w]

    def close(self) -> None:
        if self._cap is not None:
            self._cap.release()
            self._cap = None

File: go-cavebot/internal/vision/capture.go
package vision

import (
	"image"

	"gocv.io/x/gocv"
)

// FrameCapture wraps GoCV VideoCapture for reading OBS Virtual Camera frames.
type FrameCapture struct {
	cameraIndex int
	cap         *gocv.VideoCapture
}

func NewFrameCapture(cameraIndex int) *FrameCapture {
	return &FrameCapture{cameraIndex: cameraIndex}
}

func (fc *FrameCapture) Open() bool {
	cap, err := gocv.OpenVideoCapture(fc.cameraIndex)
	if err != nil {
		return false
	}
	fc.cap = cap
	return fc.cap.IsOpened()
}

func (fc *FrameCapture) Read(dst *gocv.Mat) bool {
	if fc.cap == nil {
		return false
	}
	return fc.cap.Read(dst)
}

func (fc *FrameCapture) Close() {
	if fc.cap != nil {
		fc.cap.Close()
		fc.cap = nil
	}
}

// CropRegion extracts a sub-region from a frame. region is [x, y, w, h].
func CropRegion(frame gocv.Mat, region [4]int) gocv.Mat {
	x, y, w, h := region[0], region[1], region[2], region[3]
	rect := image.Rect(x, y, x+w, y+h)
	return frame.Region(rect)
}
